stdlib/domain.ml

```diff
@@ -187,25 +211,45 @@ let spawn f =
   let term_condition = Condition.create () in
   let term_state = ref Running in
 
-  let at_startup = Atomic.get startup_function in
+  let at_each_spawn = Atomic.get startup_function in
   let body () =
     let result =
       match
         DLS.create_dls ();
         DLS.set_initial_keys pk;
-        at_startup ();
-        f ()
+        at_each_spawn ();
+        let res = f () in
+        res
       with
       | x -> Ok x
       | exception ex -> Error ex
     in
-    do_at_exit ();
+
+    let result' =
+      (* Run the [at_exit] callbacks when the domain computation either
+         terminates normally or exceptionally. *)
+      match do_at_exit () with
+      | () -> result
+      | exception ex ->
+          begin match result with
+          | Ok _ ->
+              (* If the domain computation terminated normally, but the
+                 [at_exit] callbacks raised an exception, then return the
+                 exception. *)
+              Error ex
+          | Error _ ->
+              (* If both the domain computation and the [at_exit] callbacks
+                 raised exceptions, then ignore the exception from the
+                 [at_exit] callbacks and return the original exception. *)
+              result
+          end
+    in
```

Comment:
This is a lot of code for not-clearly-desirable behavior, but I don't know of a simple way to do much shorter, and the idea of following `at_exit` is right.

Reply:
Indeed. Hopefully, we can clear this up when we find a neat solution for unhandled exceptions: https://github.com/ocaml/ocaml/issues/11074.